File: 2021/day10.py
from os.path import isfile, join as path_join
from sys import path as sys_path
sys_path.insert(1, path_join(sys_path[0], '..'))
file_name = path_join('input', 'day10.txt')
def to_list(mf: object = int, sep='\n'): return [mf(x) for x in open(file_name).read().split(sep) if x]
def to_gen(mf: object = int, sep='\n'): return (mf(x) for x in open(file_name).read().split(sep) if x)
def p1(*args): ans = part1(*args); print(ans); return ans
def p2(*args): ans = part2(*args); print(ans); return ans

if not isfile(file_name):
	from aoc import get_input_file
	get_input_file(session_path=['..', '.env'])

from utils import parsefile

data = parsefile(file_name, ["\n"])

score_corrupt = {")":3, "]":57, "}":1197, ">":25137}
score_closing = {")":1, "]":2, "}":3, ">":4}

open_c = set(["(", "[", "{", "<"])
close_c = set([")", "]", "}", ">"])

match_open = {")":"(", "]":"[", "}":"{", ">":"<"}
match_close = {"(":")", "[":"]", "{":"}", "<":">"}

def check_corrupted(line):
	stack = []
	
	for c in line:
		if c in open_c:
			stack.append(c)
		else:
			if stack[-1] == match_open[c]:
				stack.pop()
			else:
				return True, c
	
	return False, ""

non_corrupt = []

def part1():
	s = 0
	for line in data:
		res = check_corrupted(line)
		if res[0]:
			s += score_corrupt[res[1]]
		else:
			non_corrupt.append(line)

	return s


def check_closing(line):
	stack = []
	
	for c in line:
		if c in open_c:
			stack.append(c)
		else:
			if stack[-1] == match_open[c]:
				stack.pop()
			else:
				pass
	
	if len(stack) > 0:
		return False, stack
	return True, []

def calc_score(ending):
	s = 0
	for c in ending:
		s *= 5
		s += score_closing[c]
		
	return s


def part2():
	s = []
	
	for line in non_corrupt:
		res = check_closing(line)
		if not res[0]:
			ending = list(map(lambda x: match_close[x], list(res[1][::-1])))
			s.append(calc_score(ending))
	
	s = sorted(s)
	mid_i = len(s) // 2
	return s[mid_i]

p1()
p2()

File: cpp/2021/day10.cpp
#include "../aocHelper.h"

class Day10 : public BaseDay
{
public:
	Day10() : BaseDay("10") {}

	result_type solve() override
	{
		long long part1 = 0, part2 = 0;

		const unordered_map<char, int> score_corrupt = {
			{ ')', 3 },
			{ ']', 57 },
			{ '}', 1197 },
			{ '>', 25137 },
		};

		const unordered_map<char, int> score_closing = {
			{ ')', 1 },
			{ ']', 2 },
			{ '}', 3 },
			{ '>', 4 },
		};

		const unordered_set<char> open_set = { '(', '[', '{', '<' };
		const unordered_set<char> close_set = { ')', ']', '}', '>' };

		const unordered_map<char, char> match_open = {
			{ ')', '(' },
			{ ']', '[' },
			{ '}', '{' },
			{ '>', '<' },
		};

		const unordered_map<char, char> match_close = {
			{ '(', ')' },
			{ '[', ']' },
			{ '{', '}' },
			{ '<', '>' },
		};

		vector<string> lines;

		string s;

		while (*input != '\0')
		{
			if (*input == '\n')
			{
				lines.push_back(s);
				s = "";
			}
			else
			{
				s += *input;
			}
			input++;
		}

		// part 1
		auto check_corrupted = [&](string& line)
		{
			deque<char> stack;

			for (auto c : line)
			{
				if (open_set.contains(c))
				{
					stack.push_back(c);
				}
				else
				{
					if (stack.back() == match_open.at(c))
					{
						stack.pop_back();
					}
					else
					{
						return c;
					}
				}
			}

			return '\0';
		};

		vector<string_view> non_corrupt;

		for (auto& line : lines)
		{
			char c = check_corrupted(line);
			if (c != '\0')
			{
				part1 += score_corrupt.at(c);
			}
			else
			{
				non_corrupt.push_back(line);
			}
		}

		// part 2
		auto check_closing = [&](string_view& line)
		{
			deque<char> stack;

			for (auto c : line)
			{
				if (open_set.contains(c))
				{
					stack.push_back(c);
				}
				else
				{
					if (stack.back() == match_open.at(c))
					{
						stack.pop_back();
					}
					else
					{
						continue;
					}
				}
			}

			return stack;
		};

		auto calc_score = [&](deque<char>& ending)
		{
			long long s = 0;

			for (auto c : ending)
			{
				s *= 5;
				s += score_closing.at(c);
			}

			return s;
		};

		vector<long long> scores;

		for (auto& line : non_corrupt)
		{
			deque<char> stack = check_closing(line);
			if (stack.size() > 0)
			{
				reverse(stack.begin(), stack.end());
				for_each(stack.begin(), stack.end(), [&](char& c)
				{
					c = match_close.at(c);
				});

				scores.push_back(calc_score(stack));
			}
		}

		sort(scores.begin(), scores.end());
		int mid_i = scores.size() / 2;
		part2 = scores[mid_i];

		return { part1, part2 };
	}
};
